src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+        /// </summary>
+        private async Task<Solution> RewriteSolutionAsync(Document originalDocument,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites,
+            bool allOccurrences, string parameterName, CancellationToken cancellationToken)
+        {
+            // Need a copy of the expression to use to find the original expression
+            // because it gets modified when tracking the nodes.
+            var expressionCopy = expression;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            // Need to track the nodes in the expression so that they can be found later.
+            expression = expression.TrackNodes(mappingDictionary.Keys);
+            var identifiers = expression.DescendantNodes().Where(node => node is TIdentifierNameSyntax);
+
+            var modifiedSolution = originalDocument.Project.Solution;
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    modifiedSolution = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                        originalDocument, document, modifiedSolution, methodSymbol, containingMethod, expressionCopy,
+                        expression, allOccurrences, parameterName, invocations, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter.
+        /// </summary>
+        public static async Task<Dictionary<TIdentifierNameSyntax, IParameterSymbol>> MapExpressionToParametersAsync(
+            Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
```

Comment:
i feel like i haven't seen the code that bails out if the expression references something like a local symbol.  where is that, and do you test for that?

Reply:
the code that bails out based on that is in ShouldExpressionDisplayCodeActionAsync. There is a test for that called TestSimpleExpressionCaseWithLocal